### PR TITLE
Record line coverage for tia

### PR DESCRIPTION
**Pull request by DreierF**
- [ ] [No new Teamscale findings](https://build.cqse.eu/teamscale/delta.html#input/jacoco-client/?showMergeFindings=true)
- [ ] [Changes are tested adequately](https://build.cqse.eu/teamscale/tests.html#/jacoco-client/)
- [ ] USAGE.md updated in case of user-visible changes
- [ ] CHANGELOG.md updated
- [ ] build.gradle version number increased according to semantic versioning

### CODE REVIEW

File: report-generator/src/main/java/com/teamscale/report/testwise/jacoco/analysis/CachingClassAnalyzer.java [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
-import org.jacoco.core.internal.flow.ClassProbesVisitor;
+import org.jacoco.core.internal.analysis.ClassAnalyzer;
+import org.jacoco.core.internal.analysis.ClassCoverageImpl;
+import org.jacoco.core.internal.analysis.StringPool;
+import org.jacoco.core.internal.analysis.filter.Filters;
 import org.jacoco.core.internal.flow.MethodProbesVisitor;
+import org.objectweb.asm.AnnotationVisitor;
+
+import java.util.HashSet;
+import java.util.Set;
```
Comment (albertsteckermeier) [inline report-generator/src/main/java/com/teamscale/report/testwise/jacoco/analysis/CachingClassAnalyzer.java]: Unused import statements.

File: report-generator/src/main/java/com/teamscale/report/testwise/jacoco/analysis/CachingClassAnalyzer.java [hunk 2]
```diff
@@ -22,20 +26,26 @@
...
 	public void visitSource(String source, String debug) {
+		super.visitSource(source, debug);
 		classCoverageLookup.setSourceFileName(source);
 	}
 
 	@Override
 	public MethodProbesVisitor visitMethod(final int access, final String name,
 										   final String desc, final String signature, final String[] exceptions) {
-		return new CachingMethodAnalyzer(classCoverageLookup);
+		return new CachingMethodAnalyzer(classCoverageLookup, Filters.ALL, this);
```
Comment (albertsteckermeier) [inline report-generator/src/main/java/com/teamscale/report/testwise/jacoco/analysis/CachingClassAnalyzer.java]: Base class overrides visitEnd of MethodAnalyzer and modifies the base class coverage field. Do we no longer need this behavior?

File: report-generator/src/main/java/com/teamscale/report/testwise/jacoco/cache/AnalyzerCache.java [hunk 3]
```diff
@@ -1,8 +1,12 @@
 package com.teamscale.report.testwise.jacoco.cache;
 
-import com.teamscale.report.testwise.jacoco.analysis.CachingClassAnalyzer;
 import com.teamscale.report.jacoco.FilteringAnalyzer;
+import com.teamscale.report.testwise.jacoco.analysis.CachingClassAnalyzer;
 import com.teamscale.report.util.ILogger;
+import org.jacoco.core.data.ExecutionData;
+import org.jacoco.core.internal.analysis.ClassAnalyzer;
```
Comment (albertsteckermeier) [inline report-generator/src/main/java/com/teamscale/report/testwise/jacoco/cache/AnalyzerCache.java]: Unused imports.

File: report-generator/src/main/java/com/teamscale/report/testwise/jacoco/cache/ClassCoverageLookup.java [hunk 4]
```diff
@@ -31,15 +31,7 @@
 	 * List of method's line ranges. The index in this list corresponds to the probe ID.
 	 * So the same {@link LineRange}s can appear multiple times in the list if a method contains more than one probe.
```
Comment (albertsteckermeier) [inline report-generator/src/main/java/com/teamscale/report/testwise/jacoco/cache/ClassCoverageLookup.java]: Comment still talks about LineRanges which doesn't match field anymore.

File: report-generator/src/main/java/com/teamscale/report/testwise/model/builder/FileCoverageBuilder.java [hunk 5]
```diff
@@ -19,7 +22,7 @@
 	private final String fileName;
 
 	/** A list of line ranges that have been covered. */
```
Comment (albertsteckermeier) [inline report-generator/src/main/java/com/teamscale/report/testwise/model/builder/FileCoverageBuilder.java]: A set of line numbers.

File: report-generator/src/main/java/com/teamscale/report/testwise/model/builder/FileCoverageBuilder.java [hunk 6]
```diff
@@ -39,73 +42,69 @@ public String getPath() {
...
+		coveredLines.add(line);
 	}
 
 	/** Adds a line range as covered. */
 	public void addLineRange(int start, int end) {
-		coveredRanges.add(new LineRange(start, end));
+		coveredLines.addAll(IntStream.rangeClosed(start, end).boxed().collect(Collectors.toList()));
 	}
 
 	/** Adds a line range as covered. */
```
Comment (albertsteckermeier) [inline report-generator/src/main/java/com/teamscale/report/testwise/model/builder/FileCoverageBuilder.java]: Adds set of lines as covered. Please also rename the method.

File: report-generator/src/main/java/com/teamscale/report/testwise/model/LineRange.java [hunk 7]
```diff
@@ -3,37 +3,18 @@
...
-		if (start > line || start == UNKNOWN_LINE) {
-			start = line;
-		}
-		if (end < line || end == UNKNOWN_LINE) {
-			end = line;
-		}
-	}
-
 	/** @see #start */
 	public int getStart() {
```
Comment (albertsteckermeier) [inline report-generator/src/main/java/com/teamscale/report/testwise/model/LineRange.java]: Method unused.

File: report-generator/src/main/java/com/teamscale/report/testwise/jacoco/analysis/CachingClassAnalyzer.java (hunk 2 above)
Comment (DreierF) [reply report-generator/src/main/java/com/teamscale/report/testwise/jacoco/analysis/CachingClassAnalyzer.java]: No we never had it. It appends the coverage produced by the analyzed method to the ClassCoverageImpl object. We store all the info we need in ClassCoverageLookup instead.

**Pull request merged**